features/323-login-id/design.md

```diff
@@ -0,0 +1,184 @@
+- API consumer could specify realm along with login ID when realm is required.
+
+This solution should support majority of use case for compound login ID.
+
+For example, when logging in with realms:
+```json
+{ "realm": "teacher", "login_id": "a@example.com", "password": "12345678" }
+{ "realm": "student", "login_id": "a@example.com", "password": "12345678" }
+```
+If authentication succeed, both request would result in same user, with
+different principals.
+
+This addresses use case 10.
+
+## Allow Multiple Login ID per Login ID Key
+In some website, user can associate multiple emails (potentially unlimited) with
+same user. These email may be used in auth features:
+- Login with multiple emails
+- Backup email
+- Welcome email
+
+A login ID key should allow having multiple associated login ID.
+When signing up, user should provide a list of login ID. e.g.
+```json
+{
+    "login_ids": [
+        { "username": "kiootic" },
+        { "email": "kiootic+1@oursky.com" },
+        { "email": "kiootic+2@oursky.com" }
+    ],
+    "password": "12345678"
+}
+```
+
+This addresses use case 6.
+
+
+## Standard Keys
+With reference to OpenID Connect standard claims, a list of standard keys
+would be specified:
+- `username`
+- `email`
+- `phone`
+- etc.
+
+These keys would be used in user metadata.
+
+
+## Configurable Login ID Keys
+The login ID keys are whitelisted by developer. In additional, developer can
+configure each login ID key:
+- The corresponding standard key (normalization may be applied)
+- The minimum/maximum amount of login ID
+
+The auth gear would use configured keys in password recovery and welcome message.
+
+
+## User Settings
+User settings would be added. This settings is a map of strings, and only keys
+recognized by auth gear is allowed:
+- `primary_login_id`
+
+The settings have default values, and its value can be updated later using API.
+
+This settings is intended to be consumed by auth features:
+- Welcome email/SMS: Can be configured to send to primary login ID/all supported login ID.
+
+# Design Detail
+This specification would be implemented in stages:
+1. [Implicit login ID key, Login ID realm](stage-1.md)
+2. [Multiple login IDs per key, Configurable login ID keys, Standard keys](stage-2.md)
+3. [User settings](stage-3.md)
+
+Design detail of each stage can be found in corresponding documents.
+
+# Appendix
+
+**OpenID Connect: Standard Claims**: https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
+
+## Example user object
+```diff
+    {
+        "user_id": "5bf1e4d2-e1c4-4517-93c7-7dae89261da6",
+-       "login_ids": [
+-           { "email": "kiootic@oursky.com" },
+-       ],
++       "primary_login_id": "kiootic@oursky.com",
+        "metadata": {},
+        "last_seen_at": "...",
+        "created_at": "...",
+        "created_by": "...",
+        "updated_at": "...",
+        "updated_by": "...",
+        "verified": false,
+        "verify_info": {},
+        "access_token": "..."
+    }
```

Comment:
Shall we have some data about the current session in authn response? No matter it is by password or other providers. It seems the user will need to know some information about current session? (Or not?)

Reply:
For common use cases, there should be no need to know which login ID is used to login.